sky/serve/autoscalers.py

```diff
@@ -129,96 +134,99 @@ def collect_request_information(
                 self.downscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
+                return target_num_replicas
         else:
             self.upscale_counter = self.downscale_counter = 0
-        logger.info(f'Upscale counter: {self.upscale_counter}/'
-                    f'{self.scale_up_consecutive_periods}. '
-                    f'Downscale counter: {self.downscale_counter}/'
-                    f'{self.scale_down_consecutive_periods}')
+        return self.target_num_replicas
 
-        if num_replicas_delta == 0:
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        launched_replica_infos = [
+            info for info in replica_infos if info.is_alive
+        ]
+        num_alive_replicas = len(launched_replica_infos)
+
+        self.target_num_replicas = self._get_desired_num_replicas()
+        logger.info(
+            f'Final target number of replicas: {self.target_num_replicas} '
+            f'Upscale counter: {self.upscale_counter}/'
+            f'{self.scale_up_consecutive_periods}, '
+            f'Downscale counter: {self.downscale_counter}/'
+            f'{self.scale_down_consecutive_periods} '
+            f'Number of alive replicas: {num_alive_replicas}')
+
+        scaling_options = []
+        all_replica_ids_to_scale_down: List[int] = []
+
+        def _get_replica_ids_to_scale_down(num_limit: int,) -> List[int]:
+
+            status_order = serve_state.ReplicaStatus.scale_down_decision_order()
+            launched_replica_infos_sorted = sorted(
+                launched_replica_infos,
+                key=lambda info: status_order.index(info.status)
+                if info.status in status_order else len(status_order))
+
+            return [info.replica_id for info in launched_replica_infos_sorted
+                   ][:num_limit]
+
+        if num_alive_replicas < self.target_num_replicas:
+            num_replicas_to_scale_up = (self.target_num_replicas -
+                                        num_alive_replicas)
+
+            for _ in range(num_replicas_to_scale_up):
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                       target=None))
+
+        elif num_alive_replicas > self.target_num_replicas:
+            num_replicas_to_scale_down = (num_alive_replicas -
+                                          self.target_num_replicas)
+            all_replica_ids_to_scale_down.extend(
+                _get_replica_ids_to_scale_down(
+                    num_limit=num_replicas_to_scale_down,))
+
+        for replica_id in all_replica_ids_to_scale_down:
+            scaling_options.append(
+                AutoscalerDecision(AutoscalerDecisionOperator.SCALE_DOWN,
+                                   target=replica_id))
```

Comment:
I found it more intuitive to be `AutoscalerDecision(AutoscalerDecisionOperator.SCALE_DOWN, target=all_replica_ids_to_scale_down)`.
Reason: the return value `List[AutoscalerDecision]` makes me wonder what will happen if we return a mixed list of SCALE_UP and SCALE_DOWN.

How about we return `Optional[AutoscalerDecision]` with multiple replica ids in the target?

Reply:
The initial reason for `List[AutoscalerDecision]` is that scale-up in the future should include both _x_ spot to scale up and _y_ on-demand to scale up, each with a resource override dict. 
I wonder if in the future we actually might need to return a mixed list of SCALE_UP and SCALE_DOWN, if we want to do active instance migration (e.g. terminating a more available GPU when we can get a less available GPU.)

Reply:
@Michaelvll The initial reason for `List[AutoscalerDecision]` is that scale-up in the future should include both _x_ spot to scale up and _y_ on-demand to scale up, each with a resource override dict. 
I wonder if in the future we actually might need to return a mixed list of SCALE_UP and SCALE_DOWN, if we want to do active instance migration (e.g. terminating a more available GPU when we can get a less available GPU.)

Reply:
That makes a lot sense! It would be nice to have a comment there for why we are using a list of `AutoscalerDecision` in the docstr : )